pkg/backends/calico/routes.go

```diff
@@ -15,30 +15,270 @@ package calico
 	// CALICO_STATIC_ROUTES.
-	routeString := os.Getenv("CALICO_STATIC_ROUTES")
+	ch := make(chan struct{})
+	go rg.svcInformer.Run(ch)
+	go rg.epInformer.Run(ch)
+
+	/*
+		routeString := os.Getenv("CALICO_STATIC_ROUTES")
+		cidrs := []string{}
+		for _, route := range strings.Split(routeString, ",") {
+			cidr := strings.TrimSpace(route)
+			if cidr != "" {
+				cidrs = append(cidrs, cidr)
+			}
+		}
+		rg.client.updateRoutes(cidrs)
+	*/
+	return
+}
+
+// getServiceForEndpoints retrieves the corresponding svc for the given ep
+func (rg routeGenerator) getServiceForEndpoints(ep *v1.Endpoints) (*v1.Service, string) {
+	// get key
+	key, err := cache.MetaNamespaceKeyFunc(ep)
+	if err != nil {
+		log.WithField("ep", ep.Name).WithError(err).Warn("getServiceForEndpoints: error on retrieving key for endpoint, passing")
+		return nil, ""
+	}
+	// get svc
+	svcIface, exists, err := rg.svcIndexer.GetByKey(key)
+	if err != nil {
+		log.WithField("key", key).WithError(err).Warn("getServiceForEndpoints: error on retrieving service for key, passing")
+		return nil, key
+	} else if !exists {
+		log.WithField("key", key).Warn("getServiceForEndpoints: service for key not found, passing")
+		return nil, key
+	}
+	return svcIface.(*v1.Service), key
+}
+
+// getEndpointsForService retrieves the corresponding ep for the given svc
+func (rg routeGenerator) getEndpointsForService(svc *v1.Service) (*v1.Endpoints, string) {
+	// get key
+	key, err := cache.MetaNamespaceKeyFunc(svc)
+	if err != nil {
+		log.WithField("svc", svc.Name).WithError(err).Warn("getEndpointsForService: error on retrieving key for service, passing")
+		return nil, ""
+	}
+	// get ep
+	epIface, exists, err := rg.epIndexer.GetByKey(key)
+	if err != nil {
+		log.WithField("key", key).WithError(err).Warn("getEndpointsForService: error on retrieving endpoint for key, passing")
+		return nil, key
+	} else if !exists {
+		log.WithField("key", key).Warn("getEndpointsForService: service for endpoint not found, passing")
+		return nil, key
+	}
+	return epIface.(*v1.Endpoints), key
+}
+
+// setRouteForSvc handles the main logic to check if a specified service or endpoint
+// should have its route advertised by the node running this code
+func (rg *routeGenerator) setRouteForSvc(svc *v1.Service, ep *v1.Endpoints) {
+
+	// ensure both are not nil
+	if svc == nil && ep == nil {
+		log.Warn("setRouteForSvc: both service and endpoint cannot be nil, passing...")
+		return
+	}
+
+	var key string
+	if svc == nil {
+		// ep received but svc nil
+		if svc, key = rg.getServiceForEndpoints(ep); svc == nil {
+			return
+		}
+	} else if ep == nil {
+		// svc received but ep nil
+		if ep, key = rg.getEndpointsForService(svc); ep == nil {
+			return
+		}
+	}
+
+	// see if any endpoints are on this node and advertise if so
+	// else remove the route if it also already exists
+	rg.Lock()
+	if rg.advertiseThisService(svc, ep) {
+		rg.svcRouteMap[key] = append(rg.svcRouteMap[key], svc.Spec.ClusterIP)
```

Comment:
I think this line is bugged - it always appends the ClusterIP, even if the clusterIP is already in the list.

I think we can do something simpler:

```go
rg.svcRouteMap[key] =  []string{svc.Spec.ClusterIP}
```